syz-verifier/main.go

```diff
@@ -521,15 +517,15 @@ func (srv *RPCServer) cleanup(poolIdx, vmIdx int) {
 	progs := srv.pools[poolIdx].vmRunners[vmIdx]
 	delete(srv.pools[poolIdx].vmRunners, vmIdx)
 	for _, prog := range progs {
+		if prog.done {
```

Comment:
Why do we need prog.done? And in what case can it be set to true here?
If we need it here, why don't we need this check in NextExchange?

Reply:
So, in the current architecture, we keep track of what runners have what programs through `vmRunners`, which is a map from runner index to the list of programs that have been sent to that runner. The current flaw is that we don't remove a program when we have all the results for it from `vmRunners`, only when the VM containing the Runner crashes. `done` tells us whether we've verified this program. It would be probably a good idea to change this (maybe a map from progIdx to prog per Runner) but I don't know whether I should do it in this PR or not.

Reply:
> The current flaw is that we don't remove a program when we have all the results for it from vmRunners, only when the VM containing the Runner crashes.

Couldn't we simply delete the program from vmRunners in NextExchange? We have pool id, vm id and program id in NextExchange, we received the result, so the VM is definitely not running this program anymore.

> done tells us whether we've verified this program.

But whenever we set prog.done we also remove the program from srv.progs, so we will simply not find it second time in srv.progs, no? I am still not sure how we can find the program with done=true.

Reply:
> Couldn't we simply delete the program from vmRunners in NextExchange? We have pool id, vm id and program id in NextExchange, we received the result, so the VM is definitely not running this program anymore.

Yes, we could, if we loop through the array (or create a map).

> But whenever we set prog.done we also remove the program from srv.progs, so we will simply not find it second time in srv.progs, no? I am still not sure how we can find the program with done=true.

well, we delete a reference to it, `vmRunners` will still hold a reference to it so the object is not deleted.

Reply:
Ah, I see.
Removing executed programs from vmRunners when we received the result for them looks very reasons (not keeping "dead" programs). Then I guess vmRunners needs to be a map from program id to program. Then it should be easy to delete and we won't need to deal with dead programs anymore.

Reply:
Sounds good!